packages/flutter/lib/src/widgets/layout_builder.dart

```diff
@@ -0,0 +1,205 @@
+  double getMaxIntrinsicHeight(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsValid());
+    return getIntrinsicHeight(constraints);
+  }
+
+  @override
+  bool get sizedByParent => true;
+
+  @override
+  void performResize() {
+    size = constraints.biggest;
+  }
+
+  @override
+  void performLayout() {
+    if (callback != null)
+      invokeLayoutCallback(callback);
+  }
+
+  @override
+  bool hitTestChildren(HitTestResult result, { Point position }) {
+    return child?.hitTest(result, position: position) ?? false;
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (child != null)
+      context.paintChild(child, offset);
+  }
+}
+
+class _LayoutBuilderElement extends RenderObjectElement {
+  _LayoutBuilderElement(LayoutBuilder widget) : super(widget);
+
+  @override
+  LayoutBuilder get widget => super.widget;
+
+  @override
+  _RenderLayoutBuilder get renderObject => super.renderObject;
+
+  Element _child;
+
+  @override
+  void visitChildren(ElementVisitor visitor) {
+    if (_child != null)
+      visitor(_child);
+  }
+
+  @override
+  void mount(Element parent, dynamic newSlot) {
+    super.mount(parent, newSlot);    // Creates the renderObject.
+    renderObject.callback = _layout; // The _child will be built during layout.
+  }
+
+  @override
+  void update(LayoutBuilder newWidget) {
+    super.update(newWidget);
+    assert(widget == newWidget);
+    renderObject.callback = _layout;
+    renderObject.markNeedsLayout();
+  }
+
+  @override
+  void unmount() {
+    renderObject.callback = null;
+    super.unmount();
+  }
+
+  void _layout(BoxConstraints constraints) {
+    if (widget.builder == null)
+      return;
+    owner.lockState(() {
+      // TODO(hansmuller): this code was lifted from ComponentElement.rebuild(). Refactor.
+      Widget built;
+      try {
+        built = widget.builder(this, constraints.biggest);
+        assert(() {
+          if (built == null) {
+            throw new FlutterError(
+              'A build function returned null.\n'
+              'The offending widget is: $widget\n'
+              'Build functions must never return null. '
+              'To return an empty space that causes the building widget to fill available room, return "new Container()". '
+              'To return an empty space that takes as little room as possible, return "new Container(width: 0.0, height: 0.0)".'
+            );
+          }
+          return true;
+        });
+      } catch (e, stack) {
+        _debugReportException('building $widget', e, stack);
+        built = new ErrorWidget(e);
+      }
+
+      try {
+        _child = updateChild(_child, built, null);
+        assert(_child != null);
+      } catch (e, stack) {
+        _debugReportException('building $widget', e, stack);
+        built = new ErrorWidget(e);
+        _child = updateChild(null, built, slot);
+      }
+
+      _child.renderObject.layout(constraints.loosen(), parentUsesSize: false);
```

Comment:
in particular, it shouldn't be done from inside either the lockState call or the layoutCallback call. I'm surprised it didn't trip an assertion. We shouldn't be doing layout during the layout callback, iirc...